unixfs/src/walk.rs

```diff
@@ -847,8 +851,22 @@ impl fmt::Display for Error {
     }
 }
 
+#[cfg(feature = "std")]
 impl std::error::Error for Error {}
 
+fn path_pop(path: &mut String) -> bool {
+    if path.is_empty() || path == "/" {
+        return false;
+    }
+    let truncate_idx = if let Some(last_slash_idx) = path.trim_end_matches('/').rfind('/') {
+        last_slash_idx
+    } else {
+        0
+    };
```

Comment:
I just rewrote the same full path growing on a string to build the directory trees (draft PR later today), mine ended up using `full_path.bytes().rposition(|ch| ch == b'/')`. Can you see when could that `trim_end_matches('/')` be needed in general or in this file?

Could you maybe add a direct test case for this? Can't remember a case when the path would had been just `/` however I've been looking at that (draft PR later today) so can't really remember any details here.

Agreed with [your comment](https://github.com/rs-ipfs/rust-ipfs/pull/282/files#r465363929) on the place of use that we might need to get that `poor_mans_string_paths` internal library going if no one has done it before. At least w.r.t. the unixfs paths (or dag-pb links) are also serialized as utf8 so ... Well, I guess there's the case of go (libraries?) happily serializing non-utf8 bytes as utf8. Might have to see someone report a bug on such dag-pb and then decide what to do for it. Until then utf8 paths are probably more than enough.

Now thinking this side aloud the "path" handling is always "unix" when it comes to separators. Though, separators are only present in the (draft PR later today) input and only here: the Walker API aims to provide **platform compatible tracking of paths** (so that the files can be exported). I wonder if this no_std-ification should take this to account, or should we have some intermediatery path representation which can be always transformed to the *platform paths*? I can't think of anything else than windows being the problematic one, as usual :)

Reply:
> I just rewrote the same full path growing on a string to build the directory trees (draft PR later today), mine ended up using full_path.bytes().rposition(|ch| ch == b'/'). Can you see when could that trim_end_matches('/') be needed in general or in this file?

https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.components, specifically `A trailing slash is normalized away, /a/b and /a/b/ are equivalent`.

> Now thinking this side aloud the "path" handling is always "unix" when it comes to separators. Though, separators are only present in the (draft PR later today) input and only here: the Walker API aims to provide platform compatible tracking of paths (so that the files can be exported). I wonder if this no_std-ification should take this to account, or should we have some intermediatery path representation which can be always transformed to the platform paths? I can't think of anything else than windows being the problematic one, as usual :)

The algorithm was made taking into account `utf8` and `unix` paths only. If Windows interoperability is a hard-constraint, then a standalone and multi-platform "no_std" PathBuf must first be created and it will probably take some time.